lib/configuration.js

```diff
@@ -31,58 +33,147 @@ var errors = therror.register({
+ * Configuration class to merge configuration from:
  *
- * It overrides the following nconf functions:
- * a) env() method to rename the environment variables into camelCase format.
- * b) file() to support both file and directory. If it is a directory, then it merges all the JSON files
- *    in the directory. Note that the directory is not traversed recursively.
+ *  - file(path): configuration files (if path targets a configuration file) or configuration directories
+ *    (if path targets a directory). With configuration directories, all the configuration files stored
+ *    in the directory are merged in alphabetical order. Configuration files must be either JSON files
+ *    (supporting comments with hjson) or YML files.
  *
- * NOTE: The nconf library only admits one single instance. Instantiating Configuration multiple times will
- * address the same nconf instance.
+ *  - env(whitelist): environment variables. An optional white list can be used to restrict which environment
+ *    variables are to be merged.
  *
+ *  - argv(whitelist): command-line arguments. An optional white list can be used to restrict which arguments
+ *    are to be merged.
+ *
+ * NOTE: On contrast with nconf, this module supports multiple instances.
+ *
+ * @param {Object} opts
+ *    Optional object with configuration options. Currently supported:
+ *      - delimiter. Key delimiter (by default ':' to maintain compatibility with nconf).
  * @return {Object}
- *    Configuration instance (inheriting nconf methods).
+ *    Configuration instance.
  * @constructor
  */
-function Configuration() {
+function Configuration(opts) {
+
+  var delimiter = opts && opts.delimiter || ':';
+
+  var config = {};
+
+  /**
+   * Get the configuration object.
+   *
+   * @param {String} composedKey
+   *    Optional path to the configuration object to be retrieved. By default, it retrieves an object
+   *    with the whole configuration.
+   * @return {Object}
+   *    Configuration object resulting from merging files, directories and environment variables.
+   */
+  function get(composedKey) {
+    if (!composedKey) {
+      return config;
+    }
+    var getObject = function(object, keys) {
+      var key = keys.shift();
+      if (keys.length === 0) {
+        return object[key];
+      }
+      if (typeof object[key] === 'object' && object[key] !== null) {
+        return getObject(object[key], keys);
+      }
+      return undefined;
+    };
+    return getObject(config, composedKey.split(delimiter));
+  }
+
+  /**
+   * Set a single element in the configuration.
+   *
+   * @param {String} composedKey
+   *    Path to the configuration object to be set. The delimiter (by default ':') is used to set
+   *    a value in any of the levels of the configuration object hierarchy.
+   * @param {String|Number|Array|Object} value
+   *    Value to be assigned in the configuration.
+   */
+  function set(composedKey, value) {
+    var buildObject = function(object, keys) {
```

Comment:
Minor: if `buildObject` does not access `composedKey, value` from `set`, can be removed from the closure